### ISSUE TITLE
Support 'shebang' ( #!/usr/bin/boa )

### ISSUE THREAD

**Issue by ZainlessBrombie** [boa-dev/boa#1555]
**Feature**
Support - that is ignore - the unix shebang, e.g. `#!/usr/bin/boa`
This is something nodejs and deno both do: they just ignore the unix-style shebang, and this is sometimes required to execute scripts written for either.

**Example code**

```javascript
#!/bin/whatever
console.log("I still work");
```

**Comment by jedel1043** [comment]
For anyone who wants to tackle this. The pertinent code is in:

https://github.com/boa-dev/boa/blob/08f232fe99ceec594c66e822e04d387ccfd3d6c0/boa/src/syntax/lexer/mod.rs#L106-L114

Ideally we would eagerly try to match "#!" with the lexer and, if it does match, jump to the next line. You can see an example on how to do this in the `SingleLineComment` lexer:

https://github.com/boa-dev/boa/blob/8f590d781a38e0623993abd17989292d1ef61c59/boa/src/syntax/lexer/comment.rs#L23-L47

However, if implementing the logic inside `new` is too complex or too unreadable, then we can lex it on:

https://github.com/boa-dev/boa/blob/08f232fe99ceec594c66e822e04d387ccfd3d6c0/boa/src/syntax/lexer/mod.rs#L176-L286

matching for "#!" but only if `self.cursos.pos()` is in the line 1, column 1.

**Comment by Razican** [comment]
I don't think we should change the parser or the lexer. "shebang" is no valid JS, and I don't think our engine should treat it as a "comment". Our engine could be executing JS from any input, and we must be spec compliant.

But, I see the benefit of having this in the Boa executable when providing a file as the input source. I would just change the way this executable works by checking if the first line is a shebang. If it is, then remove that first line from the input sent to the engine.

This would make the executable compatible with this, but the engine would only accept valid JS.

**Comment by jedel1043** [comment]
> I don't think we should change the parser or the lexer. "shebang" is no valid JS, and I don't think our engine should treat it as a "comment". Our engine could be executing JS from any input, and we must be spec compliant.
> 
> But, I see the benefit of having this in the Boa executable when providing a file as the input source. I would just change the way this executable works by checking if the first line is a shebang. If it is, then remove that first line from the input sent to the engine.
> 
> This would make the executable compatible with this, but the engine would only accept valid JS.

Ah, then just skipping directly the `#!` in `eval` and `parse`? That would be much easier I think. Here:

https://github.com/boa-dev/boa/blob/26e149795a1ce0713a81fc50057ac23b58d44f79/boa/src/lib.rs#L89-L97

and here:

https://github.com/boa-dev/boa/blob/f9a82b4a13ed4398516ff2f8e9e92867a7670588/boa/src/context.rs#L804-L842

we would need to skip the first bytes of `src` if it begins with a `#!` and until we find an EOL. The downside I see is that now we're integrating parsing logic in `Context`, but the easiness of implementation would compensate for this, I think.

**Comment by Razican** [comment]
I think it's even better if we do the change only in `boa_cli`.

**Comment by jedel1043** [comment]
> I think it's even better if we do the change only in `boa_cli`.

I thought of that, but then a user downloading a js script from some page and using boa as an interpreter for some Rust application would fail, wouldn't it? I think that's not ideal.

**Comment by Razican** [comment]
> > I think it's even better if we do the change only in `boa_cli`.
> 
> I thought of that, but then a user downloading a js script from some page and using boa as an interpreter in some Rust application would fail, wouldn't it? I think that's not ideal, and I wouldn't want to force the user to manage the shebang manually when making a simple check and skip should be easy for us.

I think that is out of the scope of our engine. We have a JS engine that must only accept spec compliant strings.

Then, we provide a CLI tool that we can customize to allow for a particular use case that makes sense. If someone else wants to use our engine in their tool and allow for this, I guess they should add the specific support they need.

In the case we want to really change the engine, this should be an optional, opt-in feature.

**Comment by jedel1043** [comment]
> I think that is out of the scope of our engine. We have a JS engine that must only accept spec compliant strings.

I'd agree if it weren't for the fact that Firefox and Chrome support the shebang on their engines. Also, I think importing a module that has a shebang at the beginning would break, even if the user is just using `boa_cli`.

**Comment by Razican** [comment]
> > I think that is out of the scope of our engine. We have a JS engine that must only accept spec compliant strings.
> 
> I'd agree if it weren't for the fact that Firefox and Chrome support the shebang on their engines. Also, I think importing a module that has a shebang at the beginning would break, even if the user is just using `boa_cli`.
> 

OK, then let's add it as an optional, opt-in feature :)

**Comment by jedel1043** [comment]
Uhm, change of plans.
Hashbang support is a proposal on stage 3
https://github.com/tc39/proposal-hashbang
so we do need to lex it and parse it.

**Comment by nekevss** [comment]
I can try to take a crack at this one if it still needs to be done. Since there seemed to be some discussion, is there a preferred place to do this (boa-cli vs. lexing and parsing).

**Comment by jedel1043** [comment]
Originally we were going to implement it on boa-cli, but seeing the spec will change to support it in the language grammar, it must be done on the lexer/parser

**Comment by nekevss** [comment]
I made a Pull Request for the attempt I put together. It does catch the Hashbang off my example file from what I can tell. Not sure if there's anymore functionality that should be/needs to be built for it though. Currently, it treats the Hashbang like a TokenKind::Comment and moves the lexer onto the next token.
